### PR TITLE
Write test for create create book request

### PR DESCRIPTION
**Pull request by banguyenvareal**

### CODE REVIEW

File: app/assets/javascripts/preview_uploader.js [hunk 1]
```diff
@@ -13,4 +13,8 @@ $(function() {
   $("#image-upload").change(function(){
     readURL(this);
   });
+
+  $("#book_request_images").change(function(){
```
Comment (dinhvareal) [inline app/assets/javascripts/preview_uploader.js]: You can use `#image-upload`

File: app/controllers/book_requests_controller.rb [hunk 2]
```diff
@@ -4,5 +4,28 @@ def new
   end
 
   def create
+    @book_request = current_user.book_requests.build(book_request_params)
+    if @book_request.save
+      insert_data
```
Comment (dinhvareal) [inline app/controllers/book_requests_controller.rb]: Method's name should be describe its correct action

File: app/models/book_request_image.rb [hunk 3]
```diff
@@ -1,4 +1,5 @@
 class BookRequestImage < ApplicationRecord
   belongs_to :book_request
   validates :url, presence: true
+  mount_uploader :url, RequestImageUploader
```
Comment (dinhvareal) [inline app/models/book_request_image.rb]: ImagesUploader save image as a file, you can rename `url` to `file`

File: config/locales/vi.yml [hunk 4]
```diff
@@ -126,9 +126,27 @@ vi:
             price:
               greater_than_or_equal_to: "Giá bán không hợp lệ"
 
+  activerecord:
+    errors:
+      models:
+        book_request:
+          attributes:
+            name:
+              blank: "Nội dung của sách không được trống"
```
Comment (dinhvareal) [inline config/locales/vi.yml]: "Tiêu đề của sách không được trống"

File: config/locales/vi.yml [hunk 5]
```diff
@@ -126,9 +126,27 @@ vi:
...
+            quantity:
+              greater_than_or_equal_to: "Số lượng nhập không hợp lệ"
+              blank: "Nhập vào số lượng"
+            budget:
+              greater_than_or_equal_to: "Tầm giá  không hợp lệ"
```
Comment (dinhvareal) [inline config/locales/vi.yml]: Not use 2 spaces

File: spec/controllers/book_requests_controller_spec.rb [hunk 6]
```diff
@@ -1,3 +1,77 @@
...
+    image_files
+  end
+  let(:valid_book_request_params) do
+    {
+      name: book_request_params.name,
+      budget: book_request_params.budget,
+      quantity: book_request_params.quantity,
+      comment: book_request_params.comment,
+      book_request_images: image_files,
+      user_id: user.id
```
Comment (dinhvareal) [inline spec/controllers/book_requests_controller_spec.rb]: user_id not be in params

File: spec/controllers/book_requests_controller_spec.rb [hunk 7]
```diff
@@ -1,3 +1,77 @@
...
+    }
+  end
+  let(:invalid_book_request_params) do
+    {
+      name: book_request_params.name,
+      quantity: '1sad',
```
Comment (dinhvareal) [inline spec/controllers/book_requests_controller_spec.rb]: Only one invalid attribute is enough

File: spec/controllers/book_requests_controller_spec.rb [hunk 8]
```diff
@@ -1,3 +1,77 @@
...
+        post :create, params: { book_request: valid_book_request_params }
+      end
+      before do
+        post :create, params: { book_request: valid_book_request_params }
+      end
+      it 'total book request increase by 1' do
+        expect { subject }.to change(BookRequest, :count).by(1)
+      end
+      it 'give success flash' do
+        expect(flash[:success]).to eql(I18n.t('book_requests.create.success'))
```
Comment (dinhvareal) [inline spec/controllers/book_requests_controller_spec.rb:20]: Make sure that `:success` is unique in flash

File: app/assets/javascripts/preview_uploader.js (hunk 1 above)
Comment (banguyenvareal) [reply app/assets/javascripts/preview_uploader.js]: fixed

File: app/controllers/book_requests_controller.rb (hunk 2 above)
Comment (banguyenvareal) [reply app/controllers/book_requests_controller.rb]: fixed

**Review by dinhvareal — APPROVED** [review]

File: app/controllers/book_requests_controller.rb [hunk 9]
```diff
@@ -4,5 +4,28 @@ def new
   end
 
   def create
+    @book_request = current_user.book_requests.build(book_request_params)
+    if @book_request.save
+      upload_images
```
Comment (hieunguyentrung) [inline app/controllers/book_requests_controller.rb]: `save_images_to_book_request`

File: app/controllers/book_requests_controller.rb [hunk 10]
```diff
@@ -4,5 +4,28 @@ def new
...
+    else
+      render 'new'
+    end
+  end
+
+  private
+
+  def book_request_params
+    params.require(:book_request).permit(
+      :name, :comment, :budget, :quantity, :user_id, :book_request_images
```
Comment (hieunguyentrung) [inline app/controllers/book_requests_controller.rb]: Remove `:user_id`

File: app/controllers/book_requests_controller.rb [hunk 11]
```diff
@@ -4,5 +4,28 @@ def new
...
+    )
+  end
+
+  def upload_images
+    files = params[:book_request][:book_request_images]
```
Comment (hieunguyentrung) [inline app/controllers/book_requests_controller.rb]: `image_files`

File: app/uploaders/request_image_uploader.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+class RequestImageUploader < CarrierWave::Uploader::Base
+  storage :file
```
Comment (hieunguyentrung) [inline app/uploaders/request_image_uploader.rb:4]: Add new line below.

File: spec/controllers/book_requests_controller_spec.rb [hunk 13]
```diff
@@ -1,3 +1,70 @@
 require 'rails_helper'
 RSpec.describe BookRequestsController, type: :controller do
+  let(:book_request_params) { create(:book_request) }
```
Comment (hieunguyentrung) [inline spec/controllers/book_requests_controller_spec.rb]: If use `book_request_params `, you should use `attributes_for(:book_request)`

File: spec/controllers/book_requests_controller_spec.rb [hunk 14]
```diff
@@ -1,3 +1,70 @@
...
+  let(:valid_book_request_params) do
+    {
+      name: book_request_params.name,
+      budget: book_request_params.budget,
+      quantity: book_request_params.quantity,
+      comment: book_request_params.comment,
+      book_request_images: image_files
+    }
+  end
+  let(:invalid_book_request_params) { { name: nil } }
```
Comment (hieunguyentrung) [inline spec/controllers/book_requests_controller_spec.rb]: New line below.

File: spec/controllers/book_requests_controller_spec.rb [hunk 15]
```diff
@@ -1,3 +1,70 @@
...
+  describe '#create' do
+    include_context 'logged in'
+    context 'has valid params' do
+      subject do
+        post :create, params: { book_request: valid_book_request_params }
+      end
+      before do
```
Comment (hieunguyentrung) [inline spec/controllers/book_requests_controller_spec.rb]: Remove

File: spec/controllers/book_requests_controller_spec.rb [hunk 16]
```diff
@@ -1,3 +1,70 @@
 require 'rails_helper'
 RSpec.describe BookRequestsController, type: :controller do
+  let(:book_request_params) { create(:book_request) }
+  let(:image_files) do
+    image_files = []
+    rand(0..3).times do
```
Comment (hieunguyentrung) [inline spec/controllers/book_requests_controller_spec.rb]: `rand(1..5)`

File: spec/controllers/book_requests_controller_spec.rb [hunk 17]
```diff
@@ -1,3 +1,70 @@
...
+      end
+      it 'redirect to book request url' do
+        expect(subject).to redirect_to(BookRequest.last)
+      end
+    end
+    context 'has invalid params' do
+      subject do
+        post :create, params: { book_request: invalid_book_request_params }
+      end
+      before do
```
Comment (hieunguyentrung) [inline spec/controllers/book_requests_controller_spec.rb]: Remove.

**Review by hieunguyentrung — CHANGES_REQUESTED** [review]

File: app/controllers/book_requests_controller.rb [hunk 18]
```diff
@@ -4,5 +4,28 @@ def new
...
+
+  def book_request_params
+    params.require(:book_request).permit(
+      :name, :comment, :budget, :quantity, :book_request_images
+    )
+  end
+
+  def save_images_to_book_request
+    image_files = params[:book_request][:book_request_images]
+    image_files&.each do |file|
```
Comment (dinhvareal) [inline app/controllers/book_requests_controller.rb]: change `file` to `image_file`

File: app/uploaders/request_image_uploader.rb [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+
+  def store_dir
+    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
+  end
+
```
Comment (dinhvareal) [inline app/uploaders/request_image_uploader.rb:13]: Insert thumb version and only upload image file

**Review by dinhvareal — APPROVED** [review]

**Review by hieunguyentrung — APPROVED** [review]

**Pull request merged**